// @remotion/studio-server/dist/codemods/apply-component-visual-control.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyComponentVisualControl = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const recast = __importStar(require("recast"));
const parse_ast_1 = require("./parse-ast");
const applyComponentVisualControl = ({ file, transformation, changesMade, }) => {
    recast.types.visit(file.program, {
        visitJSXOpeningElement(path) {
            var _a;
            const { node } = path;
            if (node.name.type !== 'JSXIdentifier' ||
                node.name.name !== transformation.componentName) {
                return this.traverse(path);
            }
            if (((_a = node.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== transformation.line) {
                return this.traverse(path);
            }
            const attributes = node.attributes;
            if (!attributes) {
                return this.traverse(path);
            }
            for (const change of transformation.changes) {
                const parsed = (0, parse_ast_1.parseAst)(`a = ${(0, studio_shared_1.stringifyDefaultProps)({ props: JSON.parse(change.newValueSerialized), enumPaths: change.enumPaths })}`).program.body[0].expression.right;
                const existingAttr = attributes.find((attr) => attr.type === 'JSXAttribute' &&
                    attr.name.type === 'JSXIdentifier' &&
                    attr.name.name === change.propName);
                if (existingAttr && existingAttr.type === 'JSXAttribute') {
                    existingAttr.value = {
                        type: 'JSXExpressionContainer',
                        expression: parsed,
                    };
                    changesMade.push({
                        description: `Updated prop ${change.propName} on ${transformation.componentName}`,
                    });
                }
                else {
                    attributes.push({
                        type: 'JSXAttribute',
                        name: {
                            type: 'JSXIdentifier',
                            name: change.propName,
                        },
                        value: {
                            type: 'JSXExpressionContainer',
                            expression: parsed,
                        },
                    });
                    changesMade.push({
                        description: `Added prop ${change.propName} to ${transformation.componentName}`,
                    });
                }
            }
            return false;
        },
    });
    return { newAst: file, changesMade };
};
exports.applyComponentVisualControl = applyComponentVisualControl;
